harlequin: init at v1.0.1

purcell opened the pull request:
## Description of changes

Adds the TUI DuckDB IDE "harlequin", see https://harlequin.sh/

This is a Python program which builds with Poetry, and some additional necessary overrides have been added for poetry2nix, but also [merged upstream](https://github.com/nix-community/poetry2nix/pull/1319) so that a future sync won't break things here.

Feedback in the meantime is welcome.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

purcell committed:
harlequin: init at v1.1.1

purcell commented:
Now just waiting for the updated poetry2nix overrides to get synced across to nixpkgs.

robert-manchester reviewed (commented):
do new packages go in `pkgs/by-name/ha/harlequin` or is that a failed initiative?

purcell commented:
> do new packages go in `pkgs/by-name/ha/harlequin` or is that a failed initiative?

Unsure, but happy to move if that's the case.

purcell commented on pkgs/applications/misc/harlequin/update.sh:
```diff
@@ -0,0 +1,16 @@
+
+owner=tconbeer
+repo=harlequin
+latestVersion=$(curl "https://api.github.com/repos/$owner/$repo/releases/latest" | jq -r '.tag_name' | sed 's/^v//')
```
Hehe, I'll stick with the `jq` invocation, since that formulation is common in other `update.sh` scripts.

robert-manchester commented:
> > do new packages go in `pkgs/by-name/ha/harlequin` or is that a failed initiative?
> 
> Unsure, but happy to move if that's the case.

i'm new at this so have been going over the docs and https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#quick-start-to-adding-a-package which to use [pkgs/by-name](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md) but looking over recent PRs it doesn't seem like that policy is tightly enforced.

change seems fine to me either way.

purcell commented:
I've gone ahead and moved it, and also added the poetry2nix overrides here so that we're not blocked on a sync from the upstream repo (where they have already been added). So I think I can remove the draft status from this now.

robert-manchester commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5626,6 +5626,8 @@ with pkgs;
 
   gti = callPackage ../tools/misc/gti { };
 
+  harlequin = callPackage ../by-name/ha/harlequin { };
+
```
rename `pkgs/by-name/ha/harlequin/default.nix` -> `pkgs/by-name/ha/harlequin/package.nix` and these two lines can be deleted.

purcell replied:
Nice, done now

robert-manchester reviewed (commented):
`pkgs/by-name` looks for `package.nix` rather than `default.nix` so if you rename `pkgs/by-name/ha/harlequin/default.nix` to `pkgs/by-name/ha/harlequin/package.nix` the changes to`all-packages.nix` are no longer needed.

robert-manchester commented:
> Weird. I'm guessing the failure now is because for some reason `poetry2nix` doesn't work within the `pkgs/by-name` check. The attribute builds just fine for me.

apologies for suggesting the pkgs/by-name.

i pulled your branch and looking at the workflow i think these are the commands to reproduce.
```sh
nix-shell pkgs/test/nixpkgs-check-by-name/shell.nix --run bash <<"EOF"
channel=nixos-unstable
nixpkgs=$(nix-instantiate --find-file nixpkgs -I nixpkgs=channel:"$channel")
rev=$(<"$nixpkgs"/.git-revision)
nix-build -A tests.nixpkgs-check-by-name
result/bin/nixpkgs-check-by-name .
EOF
```

not sure where to go from there other than to move the package back to where it was initially.

robert-manchester commented on pkgs/development/tools/poetry2nix/poetry2nix/overrides/default.nix:
```diff
@@ -585,7 +585,7 @@ lib.composeManyExtensions [
       duckdb = super.duckdb.overridePythonAttrs (old: {
         postPatch = lib.optionalString (!(old.src.isWheel or false)) ''
-          cd tools/pythonpkg
+          ${lib.optionalString (lib.versionOlder old.version "0.8") "cd tools/pythonpkg"}
```
I am confused by this postPatch. harlequin uses duckdb 0.8.1 and that derivation should already have postPatch set, however, when i delete this block the postPatch attribute is empty, as is the patches attribute. i wonder how it is getting cleared from the super derivation.

purcell replied:
Well, yes, but the super derivation here is the poetry2nix version of the duckdb python lib. This change aligns the override with the one in the poetry2nix repo.

purcell replied:
Basically, this doesn't use the same duckdb python package we worked on together in the other PR(s).

natsukium reviewed (changes requested):
Could you please not use poetry2nix in nixpkgs?
Please use buildPythonApplication instead.

> Packages inside nixpkgs must use the [buildPythonPackage](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonpackage-function) or [buildPythonApplication](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonapplication-function) function directly, because we can only provide security support for non-vendored dependencies.

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#how-to-contribute-a-python-package-to-nixpkgs-tools

natsukium commented on pkgs/by-name/ha/harlequin/update.sh:
Can we use `nix-update-script` instead?

purcell commented:
Thanks @natsukium, this is very good feedback. I expect it will require lots more work, since there are several new python libraries I would need to include, but I did indeed notice that `mkPoetryApplication` was not really used elsewhere in `nixpkgs`, and the reasoning for this makes sense.

purcell commented:
Switched back to a draft, and I don't know that I'm likely to work further on this. I spent a bit of time at the weekend trying to get the additional python modules packaged, but concluded it's not worth the effort for me relative to how much I want to use the software.
